Index: branches/v4.2/CRM/Core/Extensions.php
===================================================================
diff -u -N -r42789 -r42899
--- branches/v4.2/CRM/Core/Extensions.php (.../Extensions.php)  (revision 42789)
+++ branches/v4.2/CRM/Core/Extensions.php (.../Extensions.php)  (revision 42899)
@@ -123,23 +123,25 @@
         }
       }
       else {
-        $civicrmDestination = urlencode(CRM_Utils_System::url('civicrm/admin/extensions', 'reset=1'));
-        $url = CRM_Utils_System::url('civicrm/admin/setting/path', "reset=1&civicrmDestination=${civicrmDestination}");
-        CRM_Core_Session::setStatus(ts('Your extensions directory: %1 is not web server writable. Please go to the <a href="%2">path setting page</a> and correct it.<br/>',
+        if (CRM_Core_Permission::check('administer CiviCRM') && $this->isDownloadEnabled()) {
+          $civicrmDestination = urlencode(CRM_Utils_System::url('civicrm/admin/extensions', 'reset=1'));
+          $url = CRM_Utils_System::url('civicrm/admin/setting/path', "reset=1&civicrmDestination=${civicrmDestination}");
+          CRM_Core_Session::setStatus(ts('Your extensions directory: %1 is not web server writable. Please go to the <a href="%2">path setting page</a> and correct it.<br/>',
             array(
               1 => $this->_extDir,
               2 => $url,
             )
           ));
+        }
         $this->_extDir = NULL;
       }
 
-      if (!class_exists('ZipArchive')) {
+      if (!class_exists('ZipArchive') && CRM_Core_Permission::check('administer CiviCRM') && $this->isDownloadEnabled()) {
         // everyone else is dumping messages wily-nily, why can't I?
         CRM_Core_Session::setStatus(ts('You will not be able to install extensions at this time because your installation of PHP does not support ZIP archives. Please ask your system administrator to install the standard PHP-ZIP extension.'));
       }
 
-      if (empty($config->extensionsURL)) {
+      if (empty($config->extensionsURL) && CRM_Core_Permission::check('administer CiviCRM')) {
         $civicrmDestination = urlencode(CRM_Utils_System::url('civicrm/admin/extensions', 'reset=1'));
         $url = CRM_Utils_System::url('civicrm/admin/setting/url', "reset=1&civicrmDestination=${civicrmDestination}");
         CRM_Core_Session::setStatus(ts('Your Extensions Directory (%1) does not have a matching Extensions Resource URL. Please go to the <a href="%2">URL setting page</a> and correct it.<br/>',
@@ -814,5 +816,14 @@
   public function isEnabled() {
       return $this->enabled;
   }
+
+  /**
+   * Determine whether the system allows downloading new extensions
+   *
+   * @return bool
+   */
+  public function isDownloadEnabled() {
+    return (FALSE !== $this->getRepositoryUrl());
+  }
 }
